### PR TITLE
Validator client builder support

### PR DESCRIPTION
**Pull request by james-prysm**
**What type of PR is this?**

Feature
related to #10593
**What does this PR do? Why is it needed?**

This PR builds on top of https://github.com/prysmaticlabs/prysm/pull/10669

This PR provides capabilities for users to define fee recipients and gaslimit information on their corresponding validator public keys and send them to the beacon node, which will in turn call the builder APIs. These user-defined configurations are important for custom builders to distribute funds correctly and are signed by the validator key for authenticity. The information used here is similar to the prepare_beacon_proposer APIs which are sent to the execution engine, the register validator API proxies to the builder APIs.

Please refer to the corresponding PRs for additional information
https://github.com/ethereum/builder-specs/pull/2
https://github.com/ethereum/beacon-APIs/pull/209/files

**includes**

- [x] deprecates fee-recipient-config-file and url flags
- [x] introduces new validator-proposer-settings-file and url flags
- [x] updates internal naming from fee recipient to validator proposer setting
- [x] yaml config parsing support for validator proposer settings
- [x] introduce gas limit to config file
- [x] builds register validator request
- [x] calls register validator beacon API from validator client
- [x] update unit tests

**does not include**

- register validator beacon API implementation internally ( gRPC)
- register validator beacon API externally ( REST)
- caching register validator 
- dynamic gas limit ( does not calculate previous block gaslimit and compare with default gas limit or current gas limit setting)
- [gas limit range warnings](https://github.com/ethereum/builder-specs/issues/17) ( if the gas limit is too high or too low)

### CODE REVIEW

File: cmd/validator/flags/flags.go
Comment (terencechain) [inline cmd/validator/flags/flags.go:359]: We don't want the override right?

File: config/fieldparams/mainnet.go
Comment (terencechain) [inline config/fieldparams/mainnet.go:27]: I don't think field params is the place for this, same with `EthBurnAddressHex`. Field params mainly define what the field length is. What about setting it in params.BeaconConfig?

File: config/validator/service/validator-proposer-settings.go
Comment (terencechain) [inline config/validator/service/validator-proposer-settings.go:11]: I find `ValidatorProposer` a bit mouthful and expressive. Why not just `Proposer`? Examples:
- `config/validator/service/proposer-settings.go`
- `ProposerSettingsConfig`
- `ProposerOptionsConfig`

File: validator/client/builder_registration.go
Comment (terencechain) [inline validator/client/builder_registration.go:1]: Can remove here. This file is just `registration.go` in `develop` branch

File: validator/keymanager/remote/mock/mock_keymanager.go
Comment (terencechain) [inline validator/keymanager/remote/mock/mock_keymanager.go:46]: It's a little bad to hard code domain here IMO. The next person that uses it might not be happy. Couldn't you just use `s.SignatureDomain` from the parameter?

**Review by terencechain — COMMENTED** [review]
Nice work! 

I'll do a second round in depth after

File: validator/keymanager/remote/mock/mock_keymanager.go
Comment (james-prysm) [reply validator/keymanager/remote/mock/mock_keymanager.go:46]: hadn't realized this will update

File: cmd/validator/flags/flags.go
Comment (james-prysm) [reply cmd/validator/flags/flags.go:359]: updated here https://github.com/prysmaticlabs/prysm/pull/10804

File: config/validator/service/validator-proposer-settings.go
Comment (james-prysm) [reply config/validator/service/validator-proposer-settings.go:11]: sure we can call it just proposer, perhaps I should change the flag names too then

File: config/fieldparams/mainnet.go
Comment (james-prysm) [reply config/fieldparams/mainnet.go]: moved to beacon config, might want to change default ethaddress property down the road

**Comment by james-prysm** [comment]
docs added to https://github.com/prysmaticlabs/documentation/pull/451

File: config/params/config.go
Comment (terencechain) [inline config/params/config.go:196]: Comments?

File: config/validator/service/proposer-settings.go
Comment (terencechain) [inline config/validator/service/proposer-settings.go:11]: I think we can lose the suffix Config since the package name already has config.  Maybe `ProposerSettingsCli`?

File: validator/client/runner.go
Comment (james-prysm) [reply validator/client/runner.go:186]: if for some reason it doesn't update correctly due to some sort of connection issue that epoch i was thinking it could just be updated in the next epoch. i was afraid of erroring out the whole system here. what do you think?

File: validator/node/node.go
Comment (james-prysm) [reply validator/node/node.go:580]: I believe by default when unmarshalling the YAML or JSON and it's empty it will set it to 0 so we should just default it if that's the case.

File: validator/node/node.go
Comment (james-prysm) [reply validator/node/node.go:583]: https://github.com/prysmaticlabs/prysm/issues/10810 issue created here

File: validator/node/node.go
Comment (james-prysm) [reply validator/node/node.go:476]: https://github.com/prysmaticlabs/prysm/issues/10809 issue created here

File: config/validator/service/proposer-settings.go
Comment (james-prysm) [reply config/validator/service/proposer-settings.go:11]: trying out naming it payload instead, let me know what you think

File: config/validator/service/proposer-settings.go
Comment (james-prysm) [reply config/validator/service/proposer-settings.go:19]: renamed to ProposerOptionPayload

File: config/params/config.go
Comment (james-prysm) [reply config/params/config.go:196]: adding comments

File: config/validator/service/proposer-settings.go
Comment (james-prysm) [reply config/validator/service/proposer-settings.go]: I'll try to have a better way of doing this using functions to cast and remove the intermediate type all together in a later PR.

File: validator/node/node.go
Comment (terencechain) [reply validator/node/node.go:581]: Forgot the unmarshal case. Thanks!

But hhy do you need the second condition? `|| gasLimit == params.BeaconConfig().DefaultBuilderGasLimit`

File: validator/client/runner.go
Comment (terencechain) [reply validator/client/runner.go:186]: Yeah. I see both sides. But if it's not intermittent then it'd always be warning. I'm ok leaving it as it is

File: beacon-chain/blockchain/execution_engine.go
Comment (james-prysm) [reply beacon-chain/blockchain/execution_engine.go:275]: I was thinking about this as well but heard that there could be other addresses as the burn address ( not sure if totally correct). open to do either or.

File: validator/client/runner.go
Comment (james-prysm) [reply validator/client/runner.go:249]: I believe this is just tech debt, it's an interface from before my time.

File: validator/client/validator.go
Comment (james-prysm) [reply validator/client/validator.go:980]: when it runs per epoch I have it running in its own go routine you can see that in the runner.go file line 190. otherwise it runs when status updates and on startup which are not on their on go routine. 

I think those are all possible future improvements, with it being set every6.4 minutes and on startup as well I wonder if that's 100% needed at this moment. I can make a note on the tracking of this feature but I don't think it should be part of this particular PR.

File: validator/node/testdata/test-unmarshal-good.yaml
Comment (james-prysm) [reply validator/node/testdata/test-unmarshal-good.yaml:3]: i just have it as a simple test for those functions

File: validator/node/node.go
Comment (james-prysm) [reply validator/node/node.go:815]: will remove then

File: validator/node/node.go
Comment (james-prysm) [reply validator/node/node.go:581]: removed that check for now

File: validator/node/testdata/test-unmarshal-good.yaml
Comment (james-prysm) [reply validator/node/testdata/test-unmarshal-good.yaml:3]: I'll adjust it to be more appopriate

Comment (james-prysm) [reply validator/node/testdata/test-unmarshal-good.yaml]: adjusted to be more appropriate

File: validator/client/runner.go
Comment (kasey) [reply validator/client/runner.go:250]: I don't see why it would be out of scope to extend the interface to add a new method?

File: validator/client/validator.go
Comment (kasey) [reply validator/client/validator.go:980]: > with it being set every6.4 minutes and on startup as well I wonder if that's 100% needed at this moment.
I don't follow your meaning here, what is it that may not be needed based on the frequency of this running?

Are we building a prototype for integration testing, or is this something we want to be usable by end-users once this is merged? If it's a prototype then I agree that we can punt on resiliency, but I think accounting for retries and partial failures are basic requirements we need in place before this is used in production. I'll defer to @terencechain on clarifying the goal.

File: validator/client/validator.go
Comment (james-prysm) [reply validator/client/validator.go]: Updated to address this, lt me know if this works , unit test added

File: validator/client/validator.go
Comment (kasey) [inline validator/client/validator.go:993]: nitpick: it's nice to use the typed string formatters (eg `%d` for numbers, `%s` for strings) rather than `%v` where possible, because linters understand them and will complain when types are misaligned.

**Review by kasey — APPROVED** [review]

**Review by kasey — APPROVED** [review]

**Pull request merged**
